### ISSUE TITLE
Node stuck after ABCI application is restarted.

### ISSUE THREAD

**Issue by nicsor** [tendermint/tendermint#9317]
**Tendermint version**:
v0.35.9

**ABCI app**:
private app

**Environment**:
- **OS** (e.g. from /etc/os-release): Debian 11 with docker image: tendermint/tendermint:v0.35.9

**What happened**:
The ABCI application was killed / crashed on the beginBlock callback and the connection to the tendermint node was broken.

basically we end up with this call to `stateCopy, retainHeight, err = cs.blockExec.ApplyBlock` failing:
   https://github.com/tendermint/tendermint/blob/main/consensus/state.go#L1702

after the application was restarted the tendermint node remained stuck at the same height until it was restarted.

**What you expected to happen**:
I'd expect tendermint to panic at this stage instead of returning. Or at least be able to recover.
https://github.com/tendermint/tendermint/blob/main/consensus/state.go#L1712
I'd expect the same behavior as when the connection between the ABCI app and the tendermint is broken outside of this code block: tendermint is stuck until the application is back up, and continues once the connection is reestablished.

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):
1. create custom app.
2. put breakpoint on begin block callback.
3. kill app after the breakpoint is hit
4. restart the application

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:
```
2022-08-26T11:16:44Z INFO finalizing commit of block hash=AC51DABED3C1E8E06EFD94E33378697BB1FF8684E388358B5D4D8E3480E2BB44 height=144 module=consensus num_txs=0 root=723A094E6692D34F12BC70793B85430F6506EBE37FDA909EE662FE8FC4693357
2022-08-26T11:16:47Z ERROR error in proxyAppConn.BeginBlock err="rpc error: code = Unavailable desc = error reading from server: EOF" module=state
2022-08-26T11:16:47Z ERROR failed to apply block err="rpc error: code = Unavailable desc = error reading from server: EOF" height=144 module=consensus
```

**Config (you can paste only the changes you've made)**:
create empty blocks is set at 10 seconds. default configuration otherwise.

**Comment by thanethomson** [comment]
We unfortunately don't support Tendermint v0.35 any more, as per https://github.com/tendermint/tendermint/discussions/9155.

Does this happen with v0.34?

**Comment by nicsor** [comment]
Ok, good to know. So for now downgrade to 0.34 until v0.37 becomes available. I did not check on v0.34.

**Comment by nicsor** [comment]
I tried checking on v0.34.20/v0.34.21 but I could not start the network. Only thing different in the default configuration is that I am using grpc. But that's a different issue.

I was able to run: v0.34.19 which seems to have a better behavior, but there is still something that might be wrong or could be improved. The application seems to have a normal exit that is not reported to the caller. However, the node is not stuck anymore.

Thanks for your help!
